### ISSUE TITLE
bucket{le="+Inf"} and count metrics contain not equal data

### ISSUE THREAD

**Issue by misternobody** [grafana/mimir#6058]
#### Describe the bug

The metric values for request_duration_seconds_bucket{le="+Inf"} and request_duration_seconds_count differ, whereas the documentation indicates that they should be identical:

`The final bucket captures all of the measurements that were captured by the preceding bucket and any measurements that exceed the preceding bucket. The count for the +Inf <metric_name>_bucket should be equivalent to the value of the <metric_name>_count field. Both fields provide a count of all of the measurements categorized by the histogram data point as of the histogram's timestamp.`

#### To Reproduce

Compare values of 2 metrics: **<metric_name>_bucket{le="+Inf"}** and **<metric_name>_count**

For example: **sum(cortex_request_duration_seconds_bucket{le="+Inf", service="grafana-mimir-ingester", status_code="success"}) - sum(cortex_request_duration_seconds_count{status_code="success", service="grafana-mimir-ingester"}) != 0**

We did not encounter this behavior when using Victoria Metrics and Prometheus prior to migrating to the Grafana stack.

<img width="1703" alt="image" src="https://github.com/grafana/mimir/assets/1763054/d5c84384-157f-414d-9198-c8312146a267">

<img width="1685" alt="image" src="https://github.com/grafana/mimir/assets/1763054/e9ec5004-af9a-4304-9f56-65929a2d3fd2">

#### Expected behavior

The difference between `<metric_name>_bucket{le="+Inf"}` and `<metric_name>_count` values must be 0.

#### Environment

- Infrastructure: Kubernetes
- Deployment tool: Distributed Helm chart (mimir-distributed-5.1.0-weekly.248)
- Mimir Docker image: r248-c7fbeb7

#### Additional Context

There were nothing interesting in Mimir or Grafana agent logs. 

Used configuration:
```
  mimir:
    structuredConfig:
      server:
        http_server_write_timeout: 300s
        http_server_read_timeout: 300s
        http_server_idle_timeout: 300s
        grpc_server_max_recv_msg_size: 209715200
        grpc_server_max_send_msg_size: 209715200
      common:
        storage:
          backend: s3
          s3:
            region: some_aws_region
            endpoint: s3.some_aws_region.amazonaws.com
      blocks_storage:
        s3:
          bucket_name: some_aws_bucket
        bucket_store:
          max_concurrent: 10000
          index_cache:
            backend: memcached
            memcached:
              timeout: 60s
              connect_timeout: 60s
              max_idle_connections: 750
              max_async_concurrency: 100
              max_async_buffer_size: 10000000
              max_get_multi_concurrency: 750
              max_get_multi_batch_size: 1000
              max_item_size: 16777216 # 16Mb
          chunks_cache:
            backend: memcached
            memcached:
              timeout: 60s
              connect_timeout: 60s
              max_idle_connections: 750
              max_async_concurrency: 100
              max_async_buffer_size: 10000000
              max_get_multi_concurrency: 750
              max_get_multi_batch_size: 1000
              max_item_size: 16777216 # 16Mb
          metadata_cache:
            backend: memcached
            memcached:
              timeout: 60s
              connect_timeout: 60s
              max_idle_connections: 750
              max_async_concurrency: 100
              max_async_buffer_size: 10000000
              max_get_multi_concurrency: 750
              max_get_multi_batch_size: 1000
              max_item_size: 16777216 # 16Mb
      memberlist:
        cluster_label_verification_disabled: true
        cluster_label: "mimir"
      ruler_storage:
        s3:
          bucket_name: some_aws_bucket
      limits:
        ruler_max_rules_per_rule_group: 10000000
        ruler_max_rule_groups_per_tenant: 1000000
        max_global_exemplars_per_user: 1000000
        max_global_series_per_user: 0
        max_global_series_per_metric: 0
        out_of_order_time_window: 15s
        max_query_parallelism: 240
        max_cache_freshness: 10m
        max_total_query_length: 30240h
        max_fetched_chunks_per_query: 0
        ingestion_rate: 200000000
        ingestion_burst_size: 200000
        compactor_blocks_retention_period: 720h
        accept_ha_samples: true
        ha_cluster_label: cluster
        ha_replica_label: __replica__
        query_sharding_total_shards: 16
        query_sharding_max_sharded_queries: 128
        label_names_and_values_results_max_size_bytes: 419430400
        cardinality_analysis_enabled: true
        label_names_and_values_results_max_size_bytes: 512000000
      alertmanager:
        max_recv_msg_size: 209715200
        alertmanager_client:
          max_recv_msg_size: 209715200
          max_send_msg_size: 209715200
      ingester:
        ring:
          unregister_on_shutdown: false
          min_ready_duration: 120s
      store_gateway:
        sharding_ring:
          unregister_on_shutdown: false
      alertmanager_storage:
        backend: s3
        s3:
          bucket_name: some_aws_bucket
          endpoint: s3.some_aws_region.amazonaws.com
      distributor:
        max_recv_msg_size: 209715200
        ha_tracker:
          enable_ha_tracker: true
          kvstore:
            store: consul
            consul:
              host: http://consul:8500
      compactor:
        compaction_interval: 1h
        deletion_delay: 1h
      frontend:
        parallelize_shardable_queries: true
        split_queries_by_interval: 24h
        align_queries_with_step: true
        log_queries_longer_than: 10s
        max_outstanding_per_tenant: 10000
        cache_results: true
        cache_unaligned_requests: true
        grpc_client_config:
          grpc_compression: snappy
        results_cache:
          memcached:
            timeout: 60s
            connect_timeout: 60s
            max_idle_connections: 750
            max_async_concurrency: 100
            max_async_buffer_size: 10000000
            max_get_multi_concurrency: 750
            max_get_multi_batch_size: 1000
            max_item_size: 16777216
      frontend_worker:
        grpc_client_config:
          max_send_msg_size: 512000000
          grpc_compression: snappy
      query_scheduler:
        max_outstanding_requests_per_tenant: 10000
      querier:
        timeout: 280s
        max_concurrent: 16
```

The configuration of the Grafana agent is fairly standard, with minimal customizations.

**Comment by nikoncode** [comment]
We have same issue and it happens pretty frequently with mimir.

**Comment by bruksnys** [comment]
We are also encountering this issue with our mimir cluster.

**Comment by pracucci** [comment]
_This is an interesting issue and I've seen few people reporting it._

First of all, I tried to reproduce it at Grafana Labs. I've run the following query in a large Mimir cluster (the number of matched series by the query is about 3M) but I couldn't reproduce it:

```
sum by(metric_name) (label_replace({__name__=~".*_bucket",le="+Inf"}, "metric_name", "$1", "__name__", "(.+)"))
-
sum by(metric_name) (label_replace({__name__=~".*_count"}, "metric_name", "$1", "__name__", "(.+)"))
!= 0
```

So I'm asking you few questions:
- Can you consistently reproduce it or is it just random / unfrequent?
- If you can consistently reproduce it:
  - Does the issue still happen if you re-run the query after clearing the query-frontend results cache (e.g. restart all memcached replicas used by query-frontend)
  - Does the issue still happen if you disable query-sharding? Remember to clear the query-frontend cache, otherwise result may be picked up from the cache.

**Comment by misternobody** [comment]
Hey @pracucci
Thanks a lot for your reply here. RE your questions:
1. It happens sporadically, we did not find any correlation between resource consumption or Mimir metrics with this stuff. 
2. Yeah, we tried to clean up all caches but it did not help there.
3. `parallelize_shardable_queries` was disabled as far as we've tuned all values related to out of order ingestion:
```
        out_of_order_time_window: 2m
        max_cache_freshness: 2m
        ruler_evaluation_delay_duration: 2m
```
4. Right now we're on the latest stable release (`2.10.3`), and unfortunately it still happens. For example:
<img width="1692" alt="image" src="https://github.com/grafana/mimir/assets/1763054/fd4462e7-5551-4cba-8323-4117031d458c">

**Comment by misternobody** [comment]
@pracucci Here is 90 days graph for one of metrics. Some points are clearly smoothed out, but the overall picture is visible: sometimes there is a significant difference between "+inf bucket" and "count", sometimes there is none.

<img width="1684" alt="Screenshot 2023-11-07 at 11 38 51" src="https://github.com/grafana/mimir/assets/1763054/31e2b95c-e242-41e9-8752-2c9a6e70ce35">

**Comment by pracucci** [comment]
The `+Inf` and `_count` metrics are two different metrics. An easy way to explain your issue is that from time to time you hit an ingestion limit on Mimir side, so one of the two metrics (or their samples) is rejected / discarded. Can you confirm you haven't hit any max series limit (e.g. max series per tenant, max series per metric name) close to the time range when you see the difference?

**Comment by misternobody** [comment]
Indeed, I can confirm that we have either disabled limits or configured them with high values that are unlikely to be reached in our environment:
```
        ruler_max_rules_per_rule_group: 10000000
        ruler_max_rule_groups_per_tenant: 1000000
        max_global_exemplars_per_user: 1000000
        max_global_series_per_user: 0
        max_global_series_per_metric: 0
```

There is 0 value for `cortex_distributor_instance_rejected_requests_total` and `cortex_ingester_instance_rejected_requests_total` metrics:

<img width="1684" alt="Screenshot 2023-11-17 at 09 35 58" src="https://github.com/grafana/mimir/assets/1763054/9ad8e9fa-6717-41bc-8113-35507513fd9f">

Additionally, I thoroughly examined the logs for both the ingester and distributor components, and there is minimal evidence of errors. There have been no instances of reaching any limits:

<img width="1696" alt="image" src="https://github.com/grafana/mimir/assets/1763054/cbdab87a-7430-4b7d-94ad-c4fa66e72f86">

**Comment by misternobody** [comment]
Hi @pracucci, maybe any other suggestions there?

**Comment by bruksnys** [comment]
Update regarding this issue:
In our case it was due to misconfigured HA metric ingestion using Grafana agent. 
Our default scrape_interval is 30 seconds instead of 15 seconds and that causes issues with metric consistency with Mimir default values (they are optimised for scrape_interval of 15 seconds). After updating distributor ha_tracker. ha_tracker_update_timeout to 30 seconds and ha_tracker.ha_tracker_failover_timeout to 60 seconds we've noticed that newly ingested metrics have become consistent. The issue is - this took us quite a while to figure out as no other errors were reported by Mimir.

https://github.com/grafana/mimir/issues/2524 - seems to track improvements to visibility of ha deduplication and we are looking forward to improvements in this regard.
